### PR TITLE
Added wallet key pools, lookahead and improved the code structe

### PR DESCRIPTION
**Pull request by erasmospunk**

### CODE REVIEW

File: crypto/src/key/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/mod.rs:51]: We don't use `Hash` at all to avoid confusion with proper cryptography hashes. In practice, there's barely any need for this. If you need a HashSet or HashMap, please use BTreeSet or BTreeMap instead.

File: crypto/src/key/key_holder.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/key_holder.rs:28]: Same here. Remove Hash please.

File: crypto/src/key/secp256k1/mod.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/mod.rs:99]: Remove Hash

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/derivation_path.rs:60]: Not related to this PR, but I just noticed this now. Isn't it better to make the `as_vec` method an `as_slice` and return `&[ChildNumber]` instead?

File: crypto/src/key/hdkd/derivation_path.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/derivation_path.rs:155]: Please, without me having to ask for this, tests can go in two ways:

1. Exhaustive manually (something like parse_derivation_path() test below)
2. Exhaustive through randomization/propery-based testing, like most of the tests in mintlayer-core.

I'm frustrated.

Comment (TheQuantumPhysicist) [reply crypto/src/key/hdkd/derivation_path.rs:155]: To be clear, this isn't just bad because it breaks the bylaws of the code, but it also omits success/error cases + omits corner cases (empty vecs, etc).

File: wallet/types/src/account_id.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/account_id.rs:57]: Id, as per rust style guidelines.

File: wallet/types/src/keys.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/keys.rs:99]: If the only way we use this is by incrementing the value, why does this struct expose a free functionality to do anything with the number? Why not just increment it here, which guarantees the behavior to external users of this module?

File: wallet/types/src/keys.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/keys.rs:89]: Same here. There are fixed rules to how the value is set. Why don't we incorporate that logic in here instead of making this struct act like another layer of storage with set/get?

File: wallet/types/src/keys.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/keys.rs:49]: Error is ChildNumber? This is wrong. Errors should implement `std::error::Error` trait, which this does not.

File: wallet/types/src/keys.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/keys.rs:56]: Can you please explain the logic here? I'm having trouble understanding why this is like this.

File: wallet/types/src/account_id.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/account_id.rs:41]: Please add documentation to what these are

**Comment by TheQuantumPhysicist** [comment]
Looking OK. Though please check the comments. I'll do another review round later.

File: wallet/types/src/account_id.rs
Comment (TheQuantumPhysicist) [inline wallet/types/src/account_id.rs:45]: Id<Transaction> typo

File: crypto/src/key/hdkd/u31.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/hdkd/u31.rs:34]: You could call into_u32() in the From implementation below.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:166]: Since you're consuming `self` here, why not use `into_vec()` intead of a copy of that vec through the slice?

File: wallet/src/key_chain/account_key_chain.rs
Comment (TheQuantumPhysicist) [inline wallet/src/key_chain/account_key_chain.rs:38]: ConstValue?

File: wallet/src/key_chain/account_key_chain.rs
Comment (TheQuantumPhysicist) [inline wallet/src/key_chain/account_key_chain.rs:235]: Shouldn't this now be constant?

My understanding is that bitcoin core wouldn't allow this because you can't guarantee that you haven't used pre-generated addresses when you increase the gap.

File: wallet/src/key_chain/account_key_chain.rs
Comment (erasmospunk) [reply wallet/src/key_chain/account_key_chain.rs:38]: Is `ConstValue` a marker type? Because it is still possible to change the field if the struct is mut.

Comment (TheQuantumPhysicist) [reply wallet/src/key_chain/account_key_chain.rs:38]: Yes. It's for devs to realize that a value is meant to be constant over its lifetime.

File: crypto/src/key/secp256k1/extended_keys.rs
Comment (TheQuantumPhysicist) [inline crypto/src/key/secp256k1/extended_keys.rs:252]: into_vec()

File: crypto/src/lib.rs
Comment (TheQuantumPhysicist) [inline crypto/src/lib.rs:16]: extern crate is unnecessary. This is very old rust.

File: wallet/src/key_chain/master_key_chain.rs
Comment (TheQuantumPhysicist) [inline wallet/src/key_chain/master_key_chain.rs:121]: Does it make sense to make this an initialization parameter instead of just a constant?

File: wallet/src/account.rs
Comment (TheQuantumPhysicist) [inline wallet/src/account.rs:178]: Enumerating yourself is a bad idea as it's prone to off-by-one error. Usually the better solution is something like `tx.outputs().iter().enumerate()`, which will return a pair of the value and its index, which then you can use.

File: wallet/src/account.rs
Comment (TheQuantumPhysicist) [inline wallet/src/account.rs:177]: As an optimization, I would make this `WithId<Transaction>` since `get_id()` is called multiple times.

File: wallet/src/account.rs
Comment (TheQuantumPhysicist) [inline wallet/src/account.rs:180]: Do we want to assert that "something was removed"? A `debug_assert!` might help. The function `remove()` returns `Some()` when something is removed.

**Comment by TheQuantumPhysicist** [comment]
I think besides the look-ahead issue this is looking better and better.

File: wallet/src/key_chain/account_key_chain.rs
Comment (TheQuantumPhysicist) [reply wallet/src/key_chain/account_key_chain.rs:236]: But we don't handle the case of reducing the size properly, right? Why not modify that test to make the assumption that this doesn't change?

If the value won't change, then there's no need to test that it'll ever change.

**Review by TheQuantumPhysicist — APPROVED** [review]

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/leaf_key_chain.rs:148]: Here's a random idea, feel free to run with it, modify or ignore as you please.

I noticed you often have a pairs of things (here `BTreeMap` with public keys) where one of the pairs is used with `ReceiveFunds` purpose and the other is used for `Change` purpose. You could define a helper container that packages these together and gives you convenient access to the object corresponding to given purpose without having to `match` every time.

Rough outline:
```rust
struct WithPurpose<T> { // TODO: name?
    receive: T,
    change: T,
}

impl<T> WithPurpose<T> {
    // implementations obvious, omitted
    fn new(receive: T, change: T) -> Self;
    fn ref_for(&self, purpose: KeyPurpose) -> &T;
    fn mut_for(&mut self, purpose: KeyPurpose) -> &mut T;
    // Optional, useful combined with `flat_map`:
    fn into_iter(&mut self) -> impl Iterator<Item = T> + ExactSizeIterator + '_;
    fn iter(self) -> impl Iterator<Item = &T> + ExactSizeIterator;
    fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> + ExactSizeIterator + '_;
}
```

The above code could then be something like this, given `public_keys: WithPurpose<BTreeMap<_, _>>`:
```rust
let old_value = public_keys.mut_for(purpose).insert(key_index, xpub);
```

Also the return value of this whole function (`load_leaf_keys`) could be `WithPurpose<LeafKeyChain>` instead of `(LeafKeyChain, LeafKeyChain)`.

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/leaf_key_chain.rs:370]: Hmm, why `+1`?

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/leaf_key_chain.rs:368]: You can use `self.get_last_used()` and remove the dead code annotation from it.

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/leaf_key_chain.rs:259]: Again, you can use `self.get_last_issued()` and remove the dead code annotation.

**Review by iljakuklic — APPROVED** [review]
Looks good. A number of minor comments and suggestions below:

File: wallet/src/key_chain/mod.rs
Comment (pavel-kokolemin) [inline wallet/src/key_chain/mod.rs:20]: The comment says that `key_purpose` and `key_index` are hardened keys, but in the code they are normal keys.

File: wallet/src/key_chain/mod.rs
Comment (pavel-kokolemin) [inline wallet/src/key_chain/mod.rs:101]: Maybe we have a check here that the `account_index` is really hardened?

File: wallet/src/account.rs
Comment (pavel-kokolemin) [inline wallet/src/account.rs:203]: What about `DecommissionPool` outputs? I think they should be checked as well.

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (pavel-kokolemin) [inline wallet/src/key_chain/leaf_key_chain.rs:364]: Maybe it would be better to rewrite it like this?
```
        let starting_index = match self.get_last_derived_index() {
            Some(last_derived_index) => last_derived_index.increment()?.get_index(),
            None => 0,
        };
        let up_to_index = match self.usage_state.get_last_used() {
            Some(last_used) => last_used.get_index() + lookahead_size + 1,
            None => lookahead_size,
        };
```

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (pavel-kokolemin) [inline wallet/src/key_chain/leaf_key_chain.rs:463]: This may override the last used to a lower value, which is probably not a good idea.

Comment (TheQuantumPhysicist) [reply wallet/src/key_chain/leaf_key_chain.rs:463]: This is actually one of the reasons I have been against having explicit way to set_last_used instead of something that increments + checks based on the internal state.

**Review by pavel-kokolemin — APPROVED** [review]

**Review by stanislav-tkach — APPROVED** [review]

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (pavel-kokolemin) [reply wallet/src/key_chain/leaf_key_chain.rs]: I think +1 is needed here (it's now replaced by the `increment()` call). For example, if we the last used index is 5 and 20 unused addresses are needed. This will be `[6..25]` range. But because we exclude `up_to_index` when we iterate (`for i in starting_index..up_to_index`) +1 is used.

File: wallet/src/account.rs
Comment (pavel-kokolemin) [reply wallet/src/account.rs:203]: I updated it to also look at the `DecommissionPool` outputs.

File: wallet/src/account.rs
Comment (pavel-kokolemin) [reply wallet/src/account.rs:199]: Updated, thanks.

**Comment by TheQuantumPhysicist** [comment]
I think is good to go, unless you're waiting for someone else's review.

File: crypto/src/key/hdkd/derivation_path.rs
Comment (iljakuklic) [inline crypto/src/key/hdkd/derivation_path.rs:65]: Is there a problem with just returning `Some(&[])` if the paths are the same? Looks like a good neutral element with respect to path concatenation.

If not, would something like this work?
```suggestion
        self.as_slice().strip_prefix(sub_path.as_slice()).filter(|p| !p.is_empty())
```

File: wallet/src/key_chain/account_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/account_key_chain.rs:37]: Is the dead code annotation still required?

File: wallet/src/key_chain/leaf_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/leaf_key_chain.rs:282]: Nitpick: Would `ensure!` be an improvement here?

File: wallet/src/key_chain/master_key_chain.rs
Comment (iljakuklic) [inline wallet/src/key_chain/master_key_chain.rs:102]: Nitpick: I think `Itertools::exactly_one` could make this part a bit more concise.

File: wallet/types/src/keys.rs
Comment (iljakuklic) [inline wallet/types/src/keys.rs:100]: Clippy complains about unwrap here

File: wallet/types/src/keys.rs
Comment (iljakuklic) [inline wallet/types/src/keys.rs:113]: ...and here

File: wallet/types/src/keys.rs
Comment (iljakuklic) [inline wallet/types/src/keys.rs:95]: I'd prefer this to be called just `last_used` without `get_`. More consistent with (most of) rest of the code base where `get_` indicates that some work other than returning a field may be performed (e.g. `Idable::get_id` may calculate hash).

File: wallet/types/src/keys.rs
Comment (iljakuklic) [inline wallet/types/src/keys.rs:108]: Same as above regarding the `get_` prefix.

**Review by iljakuklic — APPROVED** [review]
Looks good. I left some extra comments but all quite nitpicky and safe to ignore.

**Comment by pavel-kokolemin** [comment]
> Looks good. I left some extra comments but all quite nitpicky and safe to ignore.

Thanks. I will fix these in the next PR.

**Pull request merged**
